rust/lance-encoding/src/encodings/physical/dictionary.rs

```diff
@@ -0,0 +1,410 @@
+    fn decode(
+        &self,
+        rows_to_skip: u64,
+        num_rows: u64,
+        all_null: &mut bool,
+    ) -> Result<Vec<BytesMut>> {
+        // Decode the indices
+        let indices_buffers = self
+            .indices_decoder
+            .decode(rows_to_skip, num_rows, all_null)?;
+
+        let indices_array =
+            new_primitive_array::<UInt8Type>(indices_buffers.clone(), num_rows, &DataType::UInt8);
+
+        let indices_array = indices_array.as_primitive::<UInt8Type>().clone();
+        let dictionary = self.decoded_dict.clone();
+
+        let adjusted_indices: UInt8Array = indices_array
+            .iter()
+            .map(|x| match x {
+                Some(0) => None,
+                Some(x) => Some(x - 1),
+                None => None,
+            })
+            .collect();
+
+        // Build dictionary array using indices and items
+        let dict_array =
+            DictionaryArray::<UInt8Type>::try_new(adjusted_indices, dictionary).unwrap();
+        let string_array = arrow_cast::cast(&dict_array, &DataType::Utf8).unwrap();
+        let string_array = string_array.as_any().downcast_ref::<StringArray>().unwrap();
+
+        // This is not ideal, since we go from DictionaryArray -> StringArray -> buffers (BytesMut)
+        // and later in primitive_array_from_buffers() we will go from buffers -> StringArray again.
+        // Creating the BytesMut is an unnecessary copy.
+        // But it is the best we can do in the current structure
+        let null_buffer = string_array
+            .nulls()
+            .map(|n| BytesMut::from(n.buffer().as_slice()))
+            .unwrap_or_else(BytesMut::new);
+
+        let offsets_buffer = BytesMut::from(string_array.offsets().inner().inner().as_slice());
+
+        // Empty buffer for nulls of bytes
+        let empty_buffer = BytesMut::new();
+        // let bytes_buffer = string_array.values().as_slice().into();
+        let bytes_buffer = BytesMut::from_iter(string_array.values().iter().copied());
+
+        Ok(vec![
+            null_buffer,
+            offsets_buffer,
+            empty_buffer,
+            bytes_buffer,
+        ])
+    }
+
+    fn num_buffers(&self) -> u32 {
+        self.items_decoder.num_buffers() + 2
+    }
+}
+
+#[derive(Debug)]
+pub struct DictionaryEncoder {
+    indices_encoder: Box<dyn ArrayEncoder>,
+    items_encoder: Box<dyn ArrayEncoder>,
+}
+
+impl DictionaryEncoder {
+    pub fn new(
+        indices_encoder: Box<dyn ArrayEncoder>,
+        items_encoder: Box<dyn ArrayEncoder>,
+    ) -> Self {
+        Self {
+            indices_encoder,
+            items_encoder,
+        }
+    }
+}
+
+fn get_indices_items_from_arrays(arrays: &[ArrayRef]) -> (ArrayRef, ArrayRef) {
+    let mut arr_hashmap: HashMap<&str, u8> = HashMap::new();
+    let mut curr_dict_index = 1;
+    let total_capacity = arrays.iter().map(|arr| arr.len()).sum();
+
+    let mut dict_indices = vec![0; total_capacity];
```

Comment:
Minor nit but instead of...

```
let mut some_vec = vec![0; known_length];
let mut some_vec_ctr = 0;
for ... {
  if ... {
    some_vec_ctr += 1;
  }
  some_vec[some_vec_ctr] = ...;
  some_vec_ctr += 1;
}
```

You can just do:

```
let mut some_vec = Vec::with_capacity(known_length);
for ... {
  if ... {
    some_vec.push(0);
    continue;
  }   
  some_vec.push(...);
}
```

I think the second approach is slightly more readable.